### PR TITLE
Egg Move Refactor

### LINKED ISSUE #3653: Egg Moves need a refactor
### Description

See #3562 

### Version

upcoming (Edge)

### Upcoming Version

_No response_

### Discord contact info

_No response_

### PR DESCRIPTION
**Pull request by Bassoonian**
Supersedes #4488 with a clean git history and a migration script

## Description

Refactors Egg moves so they're not a giant array with weird offsets anymore but rather structured lists like the teachable learnsets. All functionality is otherwise kept as similar to the original as possible, so this should not cause any real issues with feature branches.

`egg_moves.h` as well as the `gBaseStats` files are exactly as they would be when running the migration script on the current version of upcoming.

## Issue(s) that this PR fixes
Fixes #3653 

## **Discord contact info**
bassoonian

### CODE REVIEW

File: src/data/pokemon/egg_moves.h [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
+    MOVE_UNAVAILABLE,
+};
 
-const u16 gEggMoves[] = {
 #if P_FAMILY_BULBASAUR
-    egg_moves(BULBASAUR,
+static const u16 sBulbasaurEggMoveLearnset[] = {
         MOVE_SKULL_BASH,
         MOVE_CHARM,
         MOVE_PETAL_DANCE,
```
Comment (AlexOn1ine) [inline src/data/pokemon/egg_moves.h:11]: Could the script handle the formatting as well?

**Review by AlexOn1ine — COMMENTED** [review]
Tried running the script in my repo and got this:
```
egg_move_refactor.py:18: SyntaxWarning: invalid escape sequence '\('
  for mon in re.findall("egg_moves\((.*),", data):
egg_move_refactor.py:26: SyntaxWarning: invalid escape sequence '\('
  data = re.sub("    egg_moves\(%s," % mon, "static const u16 s%sEggMoveLearnset[] = {" % monname, data)
egg_move_refactor.py:28: SyntaxWarning: invalid escape sequence '\)'
  data = re.sub("\),\n", ",\n        MOVE_UNAVAILABLE,\n};\n", data) # add terminator to each old macro
egg_move_refactor.py:42: SyntaxWarning: invalid escape sequence '\.'
  data = re.sub("\.teachableLearnset = s%sTeachableLearnset,\n" % mon, ".teachableLearnset = s%sTeachableLearnset,\n        .eggMoveLearnset = s%sEggMoveLearnset,\n" % (mon, mon), data)
egg_move_refactor.py:44: SyntaxWarning: invalid escape sequence '\.'
  macrocheck = re.findall("\.teachableLearnset = s%sTeachableLearnset,( *)\\\\" % mon, data)
egg_move_refactor.py:46: SyntaxWarning: invalid escape sequence '\.'
  data = re.sub("\.teachableLearnset = s%sTeachableLearnset,%s\\\\" % (mon, macrocheck[0]), ".teachableLearnset = s%sTeachableLearnset,%s\\\\\n        .eggMoveLearnset = s%sEggMoveLearnset,%s\\\\" % (mon, macrocheck[0], mon, " " * (len(macrocheck[0]) + 4)), data)
```

**Comment by Bassoonian** [comment]
> Tried running the script in my repo and got this:
> 
> ```
> egg_move_refactor.py:18: SyntaxWarning: invalid escape sequence '\('
>   for mon in re.findall("egg_moves\((.*),", data):
> egg_move_refactor.py:26: SyntaxWarning: invalid escape sequence '\('
>   data = re.sub("    egg_moves\(%s," % mon, "static const u16 s%sEggMoveLearnset[] = {" % monname, data)
> egg_move_refactor.py:28: SyntaxWarning: invalid escape sequence '\)'
>   data = re.sub("\),\n", ",\n        MOVE_UNAVAILABLE,\n};\n", data) # add terminator to each old macro
> egg_move_refactor.py:42: SyntaxWarning: invalid escape sequence '\.'
>   data = re.sub("\.teachableLearnset = s%sTeachableLearnset,\n" % mon, ".teachableLearnset = s%sTeachableLearnset,\n        .eggMoveLearnset = s%sEggMoveLearnset,\n" % (mon, mon), data)
> egg_move_refactor.py:44: SyntaxWarning: invalid escape sequence '\.'
>   macrocheck = re.findall("\.teachableLearnset = s%sTeachableLearnset,( *)\\\\" % mon, data)
> egg_move_refactor.py:46: SyntaxWarning: invalid escape sequence '\.'
>   data = re.sub("\.teachableLearnset = s%sTeachableLearnset,%s\\\\" % (mon, macrocheck[0]), ".teachableLearnset = s%sTeachableLearnset,%s\\\\\n        .eggMoveLearnset = s%sEggMoveLearnset,%s\\\\" % (mon, macrocheck[0], mon, " " * (len(macrocheck[0]) + 4)), data)
> ```
> 
> Edit: it seems like the script did it's thing though

Ah, this is me not making it raw strings again. I'll fix it.

File: src/data/pokemon/egg_moves.h (hunk 1 above)
Comment (Bassoonian) [reply src/data/pokemon/egg_moves.h:11]: What do you mean with formatting?

Comment (AlexOn1ine) [reply src/data/pokemon/egg_moves.h:11]: it has 2 tabs instead of one like usually for movesets

Comment (Bassoonian) [reply src/data/pokemon/egg_moves.h:11]: Ah, that's a fair relic and a trivial fix, yeah, thanks!

**Review by AlexOn1ine — APPROVED** [review]

**Pull request merged**